### PR TITLE
Add the DistributedFusedLamb optimizer

### PR DESCRIPTION
**Pull request by sneaxiy**
### PR types
New features

### PR changes
OPs

### Describe
Add hybrid parallel DistributedFusedLamb optimizer.

- At the beginning, all workers have the whole parameters, the local gradients of the whole parameters, the partial moments.
- The local gradients of the whole parameter are reduce-scattered. Each worker have the partial reduce-scattered gradients.
- Each worker calculates the partial trust ratio div tensor using the partial reduce-scattered gradients.
- If the global norm clip is needed, each worker calculates the local gradient square L2-norm value using the partial reduce-scattered gradients and then calls `ncclAllReduce` to get the global gradient square L2-norm.
- Each worker calculates the square L2-norm value of the whole parameter.
- Each worker calculates the local square L2-norm value of the partial trust ratio div tensor, and then calls `ncclAllReduce` to get the global square L2-norm value.
- Each worker updates partial parameter, and call the `ncclAllGather` to get the whole updated parameter.

### CODE REVIEW

**Review by limin2021 — APPROVED** [review]
LGTM.

**Review by Aurelius84 — APPROVED** [review]
LGTM for dtype registerar

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 1]
```diff
@@ -297,13 +326,47 @@ def apply_gradients(self, params_grads):
         if not self._use_dynamic_loss_scaling and self._init_loss_scaling == 1.0:
             return self._optimizer.apply_gradients(params_grads)
 
+        if self._supports_check_nan_inf():
+            self._optimizer._set_scale(self._loss_scaling)
+            optimize_ops = self._optimizer.apply_gradients(params_grads)
+            found_inf = self._optimizer._found_inf
+            self._add_dynamic_loss_scaling(params_grads, found_inf)
```
Comment (zhiqiu) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:333]: Not important, but it seems params_grads is not used in `_add_dynamic_loss_scaling` in this case.

**Review by zhiqiu — APPROVED** [review]
LGTM

File: python/paddle/fluid/contrib/mixed_precision/decorator.py (hunk 1 above)
Comment (sneaxiy) [reply python/paddle/fluid/contrib/mixed_precision/decorator.py:333]: Yes. But not quite important.

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_distributed_fused_lamb_op_with_clip PROPERTIES TIMEOUT 120)
set_tests_properties(test_distributed_fused_lamb_op_without_clip PROPERTIES TIMEOUT 120)`

**Pull request merged**